bluesky-social/social-app
# [🐴] State transitions

[Reviewing without whitespace](https://github.com/bluesky-social/social-app/pull/3880/files?diff=split&w=1) helps a bit.

This PR handles state transitions from app foreground to background, including tab focus on web. It also begins to guard async data fetching from being called while another request is in-flight.

To do so, I refactored the state transition logic into a reducer. All updates to `this.status` go through this reducer. Various side effects are then called.
- state transition logic is synchronous
- consecutive state transitions don't clobber each other
  - example: open chat, then nav away while initializing, chat will resolve to correct `status`
- certain states only support a subset of actions

This should make it far more clear what's happening, and help us avoid invalid states.

### Testing
I recommend uncommenting the `debug` log within `pollEvents` for more visibility into what's happening.
- open chat, go back, should see logs setting the convo to `suspended`
- open chat, nav to a different screen (not Messages screen), convo should be `backgrounded`
- when navigating back to Messages screen, convo should be `suspended`
- on web, switch tabs and switch back, you should see logs going from `ready` -> `backgrounded` -> `ready`
- on native, background app, should see same logs
- on web, nav to different screen, then switch tabs, upon return the chat should still be in `backgrounded` state
- if you want, uncomment the various `throw` cases to test error states (some UI is needed)
- local hot reloads of `convo.ts` should also work fine
